### PR TITLE
[loader] Don't load framework Facade assemblies that are older than expected.

### PR DESCRIPTION
**Pull request by lambdageek**
If there is an older framework Facade assembly in the search order before the
one that we want, don't load it.

"older" here means just compare the major component of the version.

Fixes https://github.com/mono/mono/issues/7624

### CODE REVIEW

**Comment by lambdageek** [comment]
Is this what we want?  The rules for these assemblies where we want to override the normal strong name logic are getting pretty complicated...

**Comment by marek-safar** [comment]
I am not sure about minor (e.g. 4.0 and 4.1) but quite sure about major version mismatch

File: mono/metadata/assembly.c [hunk 1]
```diff
@@ -666,6 +666,39 @@ assembly_names_equal_flags (MonoAssemblyName *l, MonoAssemblyName *r, AssemblyNa
...
+ */
+static int
+assembly_names_compare_versions (MonoAssemblyName *l, MonoAssemblyName *r, int maxcomps)
+{
+	int i = 0;
+	if (maxcomps < 0) maxcomps = 4;
+#define CMP(field) do {				\
+		if (l-> field < r-> field && i < maxcomps) return -1;	\
+		if (l-> field > r-> field && i < maxcomps) return 1;	\
+	} while (0);
```
Comment (luhenry) [inline mono/metadata/assembly.c]: Stylistic: no need for a trailing `;`

**Comment by lambdageek** [comment]
@monojenkins backport to 2018-02

**Comment by lambdageek** [comment]
@monojenkins backport to 2017-12

**Pull request merged**